pkg/wakaama: add get set functions and cleanup client connection

leandrolanzieri opened the pull request:
### Contribution description
This adds generic get/set functions to interact with resources of the LwM2M client. Also, the client connection code is refactored to use these functions and the URI parser module.

### Testing procedure
The `examples/wakaama` application should work as usual.

### Issues/PRs references
None

leandrolanzieri commented:
I'm piggybacking the fixes to the issues spotted by the static test

PeterKietzmann reviewed (commented):
This is not a full review, just some early comments

leandrolanzieri commented on pkg/wakaama/contrib/objects/common.c:
```diff
@@ -0,0 +1,413 @@
+ */
+/**
+ * @{
+ * @ingroup     lwm2m_objects_common
```
In `common.h`:  https://github.com/RIOT-OS/RIOT/pull/16203/files#diff-de3b377f790d4240630fc961e7a5cae3c2c528f816e4951d0355e4357225a72eR11

leandrolanzieri commented on pkg/wakaama/contrib/objects/common.c:
```diff
@@ -0,0 +1,413 @@
+/*
```
What should I prefix?

leandrolanzieri commented on pkg/wakaama/Kconfig:
```diff
@@ -126,4 +126,8 @@ config LWM2M_TLSF_BUFFER
 
+config LWM2M_URI_MAX_SIZE
+    int "Maximum length of an URI allowed"
+    default 64
```
True, actually it should stay as 256

leandrolanzieri replied:
Removed this configuration option as it's actually not variable.

leandrolanzieri commented on pkg/wakaama/contrib/lwm2m_client_connection.c:
```diff
@@ -135,18 +102,11 @@ static void _set_interface(sock_udp_ep_t *ep, const netif_t *netif);
+    lwm2m_client_connection_t *new_conn;
 
-    new_conn = _connection_create(instance->id, client_data);
+    DEBUG("[lwm2m_connect_server] Connecting to server in sec. instance %d\n", sec_obj_inst_id);
```
Ok, I'll change it

leandrolanzieri commented on pkg/wakaama/contrib/lwm2m_client_connection.c:
```diff
@@ -353,31 +268,60 @@ static netif_t *_get_interface(char *host)
+        goto out;
+    }
+
+    /* if no port specified, used the default server and BS-server ports */
```
Fixed

leandrolanzieri commented on pkg/wakaama/contrib/lwm2m_client_connection.c:
```diff
@@ -353,31 +268,60 @@ static netif_t *_get_interface(char *host)
     lwm2m_client_connection_t *conn = NULL;
-    char *default_port;
-    char *host;
+    char uri[CONFIG_LWM2M_URI_MAX_SIZE];
```
Changed back to the original value

jia200x commented on pkg/wakaama/contrib/objects/common.c:
```diff
@@ -0,0 +1,413 @@
+    lwm2m_uri_t uri;
+    if (!lwm2m_stringToUri(path, path_len, &uri)) {
+        DEBUG("[lwm2m:set_resource] malformed path\n");
+        return -1;
```
```suggestion
        return -EINVAL;
```

leandrolanzieri replied:
Changed!

jia200x commented on pkg/wakaama/contrib/objects/common.c:
```diff
@@ -0,0 +1,413 @@
+    default:
+        DEBUG("[lwm2m:get_data] not supported type\n");
+        result = -1;
+        break;
```
Just in case someone else adds code right after the switch-case
```suggestion
        goto free_out;
        break;
```

leandrolanzieri replied:
I simplified the return routine, now there is only on `out` point.

jia200x commented on pkg/wakaama/contrib/objects/common.c:
```diff
@@ -0,0 +1,413 @@
+        if (data->value.asBuffer.length > out_len) {
+            DEBUG("[lwm2m:get_data] not enough space in buffer\n");
+            result = -1;
+        }
```
```suggestion
        goto free_out;
        }
```

leandrolanzieri replied:
Same for this one

jia200x reviewed (changes requested):
Some minor comments

leandrolanzieri commented on pkg/wakaama/contrib/objects/common.c:
```diff
@@ -0,0 +1,413 @@
+    case LWM2M_TYPE_OPAQUE:
+        if (data->value.asBuffer.length > out_len) {
+            DEBUG("[lwm2m:get_data] not enough space in buffer\n");
+            result = -1;
```
Changed!

leandrolanzieri commented on pkg/wakaama/contrib/objects/common.c:
```diff
@@ -0,0 +1,413 @@
+    lwm2m_object_t *object = lwm2m_get_object_by_id(client_data, uri->objectId);
+    if (!object || !object->readFunc) {
+        DEBUG("[lwm2m:get_data] could not find object with ID %d\n", uri->objectId);
+        result = -1;
```
Changed!

leandrolanzieri commented on pkg/wakaama/contrib/objects/common.c:
```diff
@@ -0,0 +1,413 @@
+    /* read the resource from the specified instance */
+    uint8_t res = object->readFunc(uri->instanceId, &data_num, &data, object);
+    if (res != COAP_205_CONTENT || data->type != expected_type) {
+        result = -1;
```
Changed!

leandrolanzieri commented on pkg/wakaama/contrib/objects/common.c:
```diff
@@ -0,0 +1,413 @@
+    lwm2m_uri_t uri;
+    if (!lwm2m_stringToUri(path, path_len, &uri)) {
+        DEBUG("[lwm2m:get_resource] malformed path\n");
+        return -1;
```
Changed!

leandrolanzieri commented on pkg/wakaama/contrib/objects/common.c:
```diff
@@ -0,0 +1,413 @@
+
+    default:
+        DEBUG("[lwm2m:get_data] not supported type\n");
+        result = -1;
```
Changed!

leandrolanzieri commented on pkg/wakaama/contrib/objects/common.c:
```diff
@@ -0,0 +1,413 @@
+    lwm2m_object_t *object = lwm2m_get_object_by_id(client_data, uri->objectId);
+    if (!object || !object->writeFunc) {
+        DEBUG("[lwm2m:get_data] could not find object with ID %d\n", uri->objectId);
+        result = -1;
```
Changed!

leandrolanzieri commented on pkg/wakaama/contrib/objects/common.c:
```diff
@@ -0,0 +1,413 @@
+
+    default:
+        DEBUG("[lwm2m:get_data] not supported type\n");
+        result = -1;
```
Changed!

leandrolanzieri commented on pkg/wakaama/contrib/objects/common.c:
```diff
@@ -0,0 +1,413 @@
+    /* write the resource of the specified instance */
+    uint8_t res = object->writeFunc(uri->instanceId, 1, data, object);
+    if (res != COAP_204_CHANGED) {
+        result = -1;
```
Changed!

leandrolanzieri commented:
@jia200x I think I addressed your comments. I updated the documentation to reflect the new possible return values. Also added checks when allocating data structure.

Finally, I rebased to resolve conflicts.

jia200x commented:
It works:

<img width="1044" alt="Captura de Pantalla 2021-07-12 a la(s) 14 04 19" src="https://user-images.githubusercontent.com/1260616/125335098-28e9d500-e31a-11eb-99ae-020e32e8049e.png">
<img width="1280" alt="Captura de Pantalla 2021-07-12 a la(s) 14 02 57" src="https://user-images.githubusercontent.com/1260616/125335106-2b4c2f00-e31a-11eb-83ea-caf2d6ce882b.png">

Also, the `lwm2m mem` command doesn't seem to show leaks.

jia200x reviewed (approved):
ACK

jia200x commented:
please squash

leandrolanzieri commented:
Squashed!

leandrolanzieri commented:
Thanks for reviewing!
